avocado/core/job.py

```diff
@@ -450,6 +460,8 @@ def cleanup(self):
                     pass
 
     def create_test_suite(self):
+        msg = "create_test_suite() is deprecated in favor of self.test_suites"
+        warnings.warn(msg, DeprecationWarning)
```

Comment:
Can you describe how `self.test_suites` will deprecate/replace `create_test_suite()`?  I understand that `self.test_suite` would be deprecated/removed in favor of `self.test_suites`, but I don't see how this data structure would directly be replacement for this method.

Summarizing my thoughts: in theory Job API users can define their own `create_test_suite()` (if they feel that brave enough), and I can still see the point in deprecating/removing that ability, but I don't think it depends on replace `test_suite` for `test_suites`.

Reply:
My idea was: `self.test_suites` will create the test suites on the fly during the first call.

Users can create test suites in two ways:

 1. Accessing `self.test_suites` (implicitly and transparant)
 2. Creating manually using `TestSuite()`'s `__init__()` method or `.from_config()` (explicitly)

IMO, the responsibility to create a TestSuite should be of this class. I understand that we should keep the API simple for the users, but again, IMO, it is intuitive and easy doing this via `TestSuite()` or `TestSuite.from_config()`.

But yes, I agree that the warning message could mention the  `TestSuite()` and `TestSuite.from_config()` methods. I will mention that on v2.

Reply:
> My idea was: `self.test_suites` will create the test suites on the fly during the first call.
> 
> Users can create test suites in two ways:
> 
> 1. Accessing `self.test_suites` (implicitly and transparant)
> 2. Creating manually using `TestSuite()`'s `__init__()` method or `.from_config()` (explicitly)
> 

I see.  It looks like you're indeed proposing a change in how Job API users can create their suite.  My opinion is that, with documented phases, users can simply override that one phase, as long as it implements the required functionality.

So, instead of:

```python

class MyJob(job):
   def __init__(self, config, test_suites):
      super(MyJob, self).__init__(config, test_suites)
      # create my custom test suites here
```

Users can do:

```python
class MyJob(job):
   def create_test_suite(self):
      # create my custom test suites here
``` 

And Job phases such as `pre_tests` would *possibly* behave more appropriately than dealing with test suites that were created during the class `__init__` .

The original idea is documented here:

 https://www.redhat.com/archives/avocado-devel/2016-September/msg00044.html

> IMO, the responsibility to create a TestSuite should be of this class. I understand that we should keep the API simple for the users, but again, IMO, it is intuitive and easy doing this via `TestSuite()` or `TestSuite.from_config()`.
> 

Yes, the `Job` class should by default, create the test suites, and the bulk of the implementation can be in the `TestSuite` class.  The point we're discussing is really about whether we consider the creation of a test suite a job phase, and give users more explicit power to override it, or not.  

For examples of how one could override it, it should be possible to have a `create_test_suite()` implementation that doesn't touch the `avocado.core.resolver` module at all, and instead creates `TestSuite` instances with `Task`s loaded from a directory with `.json` recipes (this would be an implementation for a `replay` command).

> But yes, I agree that the warning message could mention the `TestSuite()` and `TestSuite.from_config()` methods. I will mention that on v2.

Reply:
> > My idea was: `self.test_suites` will create the test suites on the fly during the first call.
> > Users can create test suites in two ways:
> > 
> > 1. Accessing `self.test_suites` (implicitly and transparant)
> > 2. Creating manually using `TestSuite()`'s `__init__()` method or `.from_config()` (explicitly)
> 
> I see. It looks like you're indeed proposing a change in how Job API users can create their suite. My opinion is that, with documented phases, users can simply override that one phase, as long as it implements the required functionality.
> 
> So, instead of:
> 
> ```python
> class MyJob(job):
>    def __init__(self, config, test_suites):
>       super(MyJob, self).__init__(config, test_suites)
>       # create my custom test suites here
> ```
> 
> Users can do:
> 
> ```python
> class MyJob(job):
>    def create_test_suite(self):
>       # create my custom test suites here
> ```
> 
> And Job phases such as `pre_tests` would _possibly_ behave more appropriately than dealing with test suites that were created during the class `__init__` .

Maybe it is not crystal clear to me what are the "custom test suites" yet. Because so far I have thought that "custom test suite" would simply be something like this: `TestSuite(custom_config, tests=custom_tests)`.

In order to create a "custom test suite", users could use the "TestSuite" API to create a "custom TestSuite".

Reply:
Right, the path to creating these "custom" test suites would most probably be by *not using* `TestSuite.from_config` and using `TestSuite(name, custom_config, tests=custom_tests)` like you suggested.

The question is where to hook this in the job... directly on `__init__()`? It seems weird to me to suggest a dunder method for something that we want to support and advertise as an API feature.

Reply:
Maybe we could extend the logic of TestSuite to the Job too:

1. Users creating Job() objects manually: we assume here that no discover logic will be executed, and this is a "custom job":

```python
custom_suite = TestSuite(custom_config, tasks=custom_tasks)
job_with_a_custom_suite = Job(config, suites=[custom_suite])
```

2. Users calling Job.from_configs(): We make this very well documented and we also assume here that users want some kind of help to build a "job". (We are going to discover the tasks from configs here using our logic)

Makes sense?

Reply:
> Maybe we could extend the logic of TestSuite to the Job too:
> 
> 1. Users creating Job() objects manually: we assume here that no discover logic will be executed, and this is a "custom job":
> 
> ```python
> custom_suite = TestSuite(custom_config, tasks=custom_tasks)
> job_with_a_custom_suite = Job(config, suites=[custom_suite])
> ```
> 

Do you mean to *also* support `suites=` or have it replace `suites_config`?  Oh, now giving a better look at the next item, I see you mean to replace it.

> 1. Users calling Job.from_configs(): We make this very well documented and we also assume here that users want some kind of help to build a "job". (We are going to discover the tasks from configs here using our logic)
> 

Just to make it clear, this would be able to receive just the job config, and optionally and additionally, test suites' config, right?

A very small nitpick from me is the naming, with `from_configs` (plural) with the user being able to provide just one config... I don't think it justifies two methods though.

> Makes sense?

I spent some time trying to understand why this looked nice, but felt (a bit) wrong to me.  One of the reasons I finally realized is that with this level of decoupling from the test suite to the job may be a limitation, I mean, preventing API users from using Job information (assuming there can be more info in the job instance in addition to the the config) when deciding how to create the test suites.  Chicken and egg kind of problem.

Given that I can not prove this will be a valid use case, I'm OK with going ahead with your proposal.  Worst case scenario, a user will never to subclass `Job` and handle the suites there too.

Reply:
I meant, "Worst case scenario, a user will **have** to subclass..."

Reply:
@clebergnu sorry if I was not clear and we taking many comments, but I'm still playing and trying to choose the best option here too. The good thing is that I can see now (by changing the current code) the possible scenarios more clear and I think that we are *almost* there.

Yes, It is clear to me that `suites` should be optional during Job() initialization. And since I'm trying to get the consistency here too, let me know if the following assumptions are good for the users too:

1.  We are going to allow both: manual test suite discover (custom by the user) and automatic test suite discover (Avocado logic);

2. We would like to have this in a **uniform** and **intuitive**/**explicit** way, in order to reduce the learning curve. This should be valid also for advanced Avocado contributors that are reading our current code. What do I mean by that? 

 * Uniform: The way to implement should be the same as for Job() and TestSuite(); (automatic discover vs manual)

 * Intuitive/explicit: If objects are created "by hand" (`Job()` or `TestSuite() `__init__()`) we are going to assume *no magic*, because users need to either pass every information needed to the constructor or add that information later. If objects are created using `.from_config()` we can apply our magic here and try to build the test suite based on the config dicts (of course that we need a very well documented method here too).

So, users will have to call `Job(config, suites)` or `job = Job(config); job.suites = suites`  when doing "something manual/custom" and `job = Job.from_config(config)` or `job = Job.from_config(config, suites_configs)` when need some help to build the TestSuite automatically. Yes, this is a little bit more code for the user, but IMO this should help things getting "intuitive and explicit".

I'm almost done with this implementation as I described, and maybe I will find this the "wrong way" in the future, but so far, looks the way to go to me.

What do you think?

Reply:
> @clebergnu sorry if I was not clear and we taking many comments, but I'm still playing and trying to choose the best option here too. The good thing is that I can see now (by changing the current code) the possible scenarios more clear and I think that we are _almost_ there.
> 
> Yes, It is clear to me that `suites` should be optional during Job() initialization. And since I'm trying to get the consistency here too, let me know if the following assumptions are good for the users too:
> 
> 1. We are going to allow both: manual test suite discover (custom by the user) and automatic test suite discover (Avocado logic);
> 2. We would like to have this in a **uniform** and **intuitive**/**explicit** way, in order to reduce the learning curve. This should be valid also for advanced Avocado contributors that are reading our current code. What do I mean by that?
> 
> * Uniform: The way to implement should be the same as for Job() and TestSuite(); (automatic discover vs manual)
> * Intuitive/explicit: If objects are created "by hand" (`Job()` or `TestSuite()` constructor) we are going to assume _no magic_, because users need to either pass every information needed to the constructor or add that information later. If objects are created using `.from_config()` we can apply our magic here and try to build the test suite based on the config dicts (of course that we need a very well documented method here too).
> 
> So, users will have to call `Job(config, suites)` or `job = Job(config); job.suites = suites` when doing "something manual/custom" and `job = Job.from_config(config)` or `job = Job.from_config(config, suites_configs)` when need some help to build the TestSuite automatically. Yes, this is a little bit more code for the user, but IMO this should help things getting "intuitive and explicit".
> 
> I'm almost done with this implementation as I described, and maybe I will find this the "wrong way" in the future, but so far, looks the way to go to me.
> 
> What do you think?

Looks very good! +1 from me.